### DISCUSSION TITLE
Support for external version file

### DISCUSSION THREAD

**Discussion by AlexGuironnetRTE** [renovatebot/renovate#10136] [Request Help] [answered]
Hello,

We've been using the Renovate bot for a while now and it's been working really well on our UI (package.json) dependencies, so I'm taking this opportunity to thank you for that ;-) The onboarding PR and default configuration made it really easy to try out.

We've just noticed however that a lot of the dependencies for the backend (managed using gradle) weren't being updated, which must be because we are setting the versions in a "[version.properties](https://github.com/opfab/operatorfabric-core/blob/develop/versions.properties)" file then getting the versions using variables in the xxx.gradle files (for example [build.gradle](https://github.com/opfab/operatorfabric-core/blob/develop/build.gradle)).

I looked in the documentation and I don't think there is any option to specify additional files where Renovate should look for dependencies, right ? If not, is there a plan to support this kind of set up ?

I think the rationale for it was to avoid repeating the versions for monorepos such as Spring Boot and risk forgetting to update some of the dependencies, but maybe now that the bot is taking care of it that's not really a problem...

Thanks in advance for any advice!

**Comment by viceice** [comment, marked as answer]
You need to use the new `gradle-lite` manager, but pay attention, we currently merging it back to `gradle` mager, see #10132 . So for now you can simply use [`:switchToGradleLite`](https://docs.renovatebot.com/presets-default/#switchtogradlelite) preset

https://github.com/renovatebot/renovate/blob/2435a7f07df1345252213c73c3efbab9089ff961/lib/manager/gradle/shallow/extract.spec.ts#L37-L39

**Reply by AlexGuironnetRTE** [reply]
Thanks a lot for the quick reply! I didn't think of looking for a Gradle-specific solution.

I added the `:switchToGradleLite` preset and added a fileMatch property to have my version.properties file taken into account as explained in the docs (see configuration [here](https://github.com/opfab/operatorfabric-core/blob/develop/renovate.json)).

I think I must have gotten this bit right as in [the renovate job lob](https://app.renovatebot.com/dashboard#github/opfab/operatorfabric-core/362361838) I see: 

    "gradle-lite": [
      {
        "packageFile": "versions.properties",
        "datasource": "maven",
        "deps": []
      },

But I assume the empty deps array shows that it didn't recognize any dependencies (and indeed I've had no PRs created for the dependencies in question even though they're outdated).

Did I miss something ?

Or could it be that the way we declare our variables in versions.properties is not recognized by renovate ?

I mean for example here is what we have in the versions.properties file :

```
# spring libs
spring.boot=2.4.0-M1
spring.cloud=Hoxton.SR9
spring=5.3.2
spring.security=5.4.2
spring.retry=1.3.0
```

I mean the bot has been very clever so far, but figuring out that we're using the version from variable "spring.boot" for dependencies "org.springframework.boot:spring-boot-gradle-plugin", "org.springframework.boot:spring-boot-starter-web" etc. might be too much to ask ?

**Reply by viceice** [reply]
@zharinov Can you please bring some more light into this? As you've mostly written the `gradle-lite` manager.

@AlexGuironnetRTE I think you also need to add all your gradle files, so renovate can find the variables and use the references from the properties. See also my linked test case above, it uses the default `filematch` 
https://github.com/renovatebot/renovate/blob/c43a4dc328bedf55777a641bc8477ff5a49a1722/lib/manager/gradle-lite/index.ts#L7-L7

**Reply by AlexGuironnetRTE** [reply]
Yes my other gradle files all follow the pattern *.gradle so they're picked by the default fileMatch, and I understood from the docs that I didn't need to repeat the pattern as the configuration works by extending the default.

I think the info under "INFO: Dependency extraction complete" in the logs confirms that they're taken into account : 

```
      "gradle-lite": {
        "fileCount": 31,
        "depCount": 19
      },
```

I've counted and we have 30 *.gradle files + versions.properties so that does make 31.

Thanks a lot for looking into it, let me know if there's anything else I can try.

**Reply by zharinov** [reply]
If I understand the problem right, we don't handle all `*.properties` files (my guess, the reason is performance), but only those that match `*.gradle.properties`.
Though it can be configured to `['(^|/)versions.properties$', '(^|/)gradle.properties$', '\\.gradle(\\.kts)?$']`, I'm not sure it will produce correct result. Let me think of it a little bit... Should not be so hard to extend our current code.

**Reply by zharinov** [reply]
Okay, now I see we're assuming that any identifier defined in `*.properties` is immediately available in the top-level scope in corresponding `*.gradle` files.

Although `versions.properties` file is parsed successfully, its values were bound to the top-level scope, not to `versions` variable. That's why strings like  `"org.springframework.boot:spring-boot-gradle-plugin:${versions['spring.boot']}",` weren't parsed. As far as I know, we don't evaluate Groovy so deeply at the moment.

**Reply by AlexGuironnetRTE** [reply]
Oh I see! 

> Okay, now I see we're assuming that any identifier defined in *.properties is immediately available in the top-level scope in corresponding *.gradle files.

From that comment I understand that what would work would be to have a XXX.properties file for each XXX.gradle file, and to reference them directly with ${spring.boot} for example. Is that correct? In this case I'll just change our build files for now, and specify versions directly in the `*.gradle` files instead of the `versions.properties` file, because several *.properties files would defeat our purpose of "having all versions in one place" I think. Thanks a lot for looking into it anyway!

Just to make sure I understood correctly (out of curiosity):
- Because the values from the versions.properties file are bound to the top-level scope, `${versions['spring.boot']}` returns an empty string so when parsing the build.gradle file, the bot is seeing `"org.springframework.boot:spring-boot-gradle-plugin:"`
- That doesn't get recognized as a dependency because the bot is looking for strings matching "XXX:XXX:XXX" not "XXX:XXX:" ?

Thanks again for the support and the great work!

**Reply by zharinov** [reply]
We ignore `${versions['spring.boot']}` at the parsing phase because interpretation of `foo[bar]` constructions is too much for simple JS-based parser. Maybe, some day we'll make powerful enough parser for things like this.

However, currently we support direct string substitutions of local variables (`def x = '1.2.3'`) as well as those from `gradle.properties` (e.g. [code examples from Gradle docs](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties)).

So yes, you can rename `versions.properties` to `gradle.properties` and use `"org.springframework.boot:spring-boot-gradle-plugin:${spring.boot}"`. The shorter form like `$spring.boot` also should work well.

**Reply by zharinov** [reply]
One more thing: driven by curiosity, you can find tricky syntactic variations of the same `"ggg:aaa:vvv"` thing that we don't support (yet), but we're aiming to cover most common ones and extend this list quickly when users demand.

**Reply by AlexGuironnetRTE** [reply]
Oh ok thanks a lot for the pointer... I can't figure out why we went with this versions.properties file that we had to load in the build file rather than the matching gradle.properties file which gets read by default... 

`So yes, you can rename versions.properties to gradle.properties and use "org.springframework.boot:spring-boot-gradle-plugin:${spring.boot}". The shorter form like $spring.boot also should work well.`

=> I've tried that but now the build doesn't run locally (I'm getting an error at the configuration stage), I suspect it is due to other weird stuff that we are doing in the build, once I manage to fix it I'll let you know how the gradle-lite manager is doing on it.

`One more thing: driven by curiosity, you can find tricky syntactic variations of the same "ggg:aaa:vvv" thing that we don't support (yet), but we're aiming to cover most common ones and extend this list quickly when users demand.`

Clearly the possibilities are endless... and in any case using renovate might solve some of the issues that are otherwise addressed with "exotic" build set-ups.

**Reply by rarkins** [reply]
@zharinov can we document - most importantly - which file names/syntax is supported for variables? And less importantly document any commonly used ones which aren't yet supported.

**Reply by AlexGuironnetRTE** [reply]
I've finally been able to switch to using `gradle.properties` instead of our homemade solution using `version.properties` and it's working great, the bot has already created 8 PRs! On a side note, I've had to change property names so they didn't have dots (spring.boot -> springBoot) otherwise Gradle would look for a `spring` variable.

Other than that, we've been using the gradle-lite manager for around 2 weeks now and everything looks fine!

Thanks again for your help!

**Reply by rarkins** [reply]
Any chance you could contribute those learnings as an informal guide in https://github.com/renovatebot/renovate/blob/main/lib/manager/gradle-lite/readme.md ? It will then be automatically published to our documentation.

**Reply by zharinov** [reply]
Yes, I remember about it

**Reply by AlexGuironnetRTE** [reply]
> Any chance you could contribute those learnings as an informal guide in https://github.com/renovatebot/renovate/blob/main/lib/manager/gradle-lite/readme.md ? It will then be automatically published to our documentation.

Sorry I wasn't sure who you were talking to, but if you mean me sure I can do that next week!

**Reply by AlexGuironnetRTE** [reply]
Ok, so not really next week sorry ;-) but I finally got around to it https://github.com/renovatebot/renovate/pull/10640
